### PR TITLE
newdevice: basic support for Ceragon devices

### PR DESCRIPTION
**Pull request by guillemmateos**
Add basic support for Ceragon devices

### CODE REVIEW

**Comment by laf** [comment]
Is ceragon the right name for the os? Vendor names usually aren't a good choice as they tend to have different products that need detecting separately.

**Comment by guillemmateos** [comment]
It's a good question @laf. I've just been checking on their website and they call it 'CeraOS'. I'll update the files to reflect that.

File: includes/polling/os/ceraos.inc.php [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?php
+
+$hardware = snmp_walk($device, 'MWRM-UNIT-MIB::genEquipInventoryCardName', '-Osqnv');
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: You only need snmp_get not walks on all of these.

File: includes/discovery/sensors/temperatures/ceraos.inc.php [hunk 2]
```diff
@@ -1,16 +1,16 @@
 <?php
 
-if ($device['os'] == 'ceragon') {
+if ($device['os'] == 'ceraos') {
     $mib  = 'MWRM-UNIT-MIB::genEquipUnitIduTemperature.0';
     $oid = ' .1.3.6.1.4.1.2281.10.1.1.9.0';
     $oids = snmp_get($device, $mib, '-OsqnU', 'MWRM-UNIT-MIB');
```
Comment (laf) [inline includes/discovery/sensors/temperatures/ceraos.inc.php]: No need to specify the MIB when you have it in $mib already.

File: includes/discovery/sensors/voltages/ceraos.inc.php [hunk 3]
```diff
@@ -1,16 +1,16 @@
 <?php
 
-if ($device['os'] == 'ceragon') {
+if ($device['os'] == 'ceraos') {
     $mib  = 'MWRM-UNIT-MIB::genEquipUnitIduVoltageInput.0';
     $oid = '.1.3.6.1.4.1.2281.10.1.1.10.0';
     $oids = snmp_get($device, $mib, '-OsqnU', 'MWRM-UNIT-MIB');
```
Comment (laf) [inline includes/discovery/sensors/voltages/ceraos.inc.php]: No need to specify the MIB when you have it in $mib already.

File: includes/discovery/sensors/temperatures/ceraos.inc.php (hunk 2 above)
Comment (guillemmateos) [inline includes/discovery/sensors/temperatures/ceraos.inc.php]: In terms of code, you prefer the $mib without the MWRM-UNIT-MIB:: or to remove it from last parameter of snmp_get?

File: includes/polling/os/ceraos.inc.php (hunk 1 above)
Comment (guillemmateos) [inline includes/polling/os/ceraos.inc.php]: Unfortunately with snmp_get it does not work. In fact when I poll the devices I have access to, the value is in genEquipInventoryCardName.127 (for example). Not sure if that 127 is always the same and hence why I used snmp_walk, but i can also do snmp_get with the .127 at the end if you prefer so.

Comment (laf) [inline includes/polling/os/ceraos.inc.php]: It most likely is so yes I'd switch to .127

**Comment by laf** [comment]
Don't forget you need to sign the contributors agreement.

**Comment by guillemmateos** [comment]
Will sign the contributors agreement also ASAP.

**Comment by laf** [comment]
bump @guillemmateos

**Comment by barryodonovan** [comment]
I hadn't seen this when I wrote my ceragon patch PR#4838

I have added multiple metrics for IP20C/E/S varients, also temp and voltage work for all IP10 and IP20.
I was just about to start graphs for IP10 when I spotted this. I also have access to IP20G to test but no IP20A or N models.

What is best way to proceed?

**Comment by barryodonovan** [comment]
Also, I agree in principle about not using a vendor name to differentiate but all ceragon radios use the same OS (cerOS) and MIBs yet they are not interoperable (and indeed different versions of the OS appear to be created for each hardware variant). I do think that using CerOS rather than ceragon is just confusing...

**Comment by guillemmateos** [comment]
As far as i'm aware all Ceragon devices as really radio devices, so they would all fit under 'ceros-radio' or 'ceros-wifi' if we're not very picky. As such, maybe just keeping it as ceros is ok.
@barryodonovan i guess IP20C/E/S all behave the same way (although right now i only have access to IP20C. I suspect the IP10 will behave differently based on the fact that it's a split mount unit?

**Comment by barryodonovan** [comment]
IP20C is dual radio in single outdoor enclosure, that's what I'm mostly used to also but we do use some other varients. IP20S is single chain Outdoor, I'm not sure about E, I've never seen one. IP20G is a 1U integrated indoor unit which uses 2 Standard ceragon outdoor RFUs, I think IP20GX allows for failover configuration. Then there's the N, A and Evolution which I think are more blade like with swappable modules. Before all that there was the IP10, mostly one IDU per RFU but allowing 1+1, 2+2, etc except for the IP10G which allowed 4 slots and 4 RFUs to one IDU! 

Anyway, history lesson aside, my PR has graphs for temp, voltage, RxLevel, TxPower, MSE, XPI, Defected blocks, TxRate and RxRate with both radios shown on a single graph for each metric.

It is primarily for IP20C currently but I am reworking it to show as many radios as are on the device found, eg single radio, dual, quad, etc.

Since you got here first I will let you decide if you want to press on with your PR and I can merge into it next week or if you want to let mine go ahead into master.

Barry

**Comment by laf** [comment]
It seems like we should just leave this as ceros.

@guillemmateos just needs OS tests doing as well. give us a shout if you need anything.

**Comment by guillemmateos** [comment]
I did include OS tests and changed to use sysObjectId. Travis CI is returning an error about build scripts which i'm not sure where it's coming from. Any hints?
Regarding CeraOS or Ceragon name, i think i do also prefer Ceragon (which is how it was initially made) but i won't push for it. Not a big deal. I think we can also change this at a later stage if desired. @barryodonovan i think you should merge all your changes as soon as we can get this pull request through as they seem really interesting. I'm also looking forward to improve support for a few devices and also improve general look of graphs if time allows.

**Comment by laf** [comment]
@barryodonovan All of our devices are based on OS rather than vendor or product name as that usually varies. That's a legacy choice but we've had no reason to change away from it so as such we need to carry on that approach.

@guillemmateos I'll take a look soon.

**Comment by barryodonovan** [comment]
@laf But they're not all based on OS. There's SAF, where you just merged a change of mine, which is also a hardware vendor who probably internally call their os SafOS.
There's mikrotik(routerOS), dozens of Cisco (iOS) devices and cambium/canopy. OS makes sense in some regards but theres also ubiquiti who have edgeOs and airOs which are actually both Linux with a micro kernel and some proprietary extensions. 

Look its your project and if you want it to be (a possibly made up name) ceros, then it will be ceros, its just clunky.

Barry.

**Comment by laf** [comment]
I didn't ask it to be ceros, I asked if ceragon was the right name and ceros was given instead. I know nothing about this vendor merely how things should be set hence I asked the question rather than saying it must be X Y or Z. If ceros is wrong it needs changing, if it's right then it doesn't.

**Comment by laf** [comment]
os doesn't just indicate the firmware a device runs, it also dictates what device type it is. So we can't have a vendor name then four different device types, they have to have different os definitions.

It's not my project by the way but it tends to need direction otherwise if we just merged the first pull request from everyone then things wouldn't work.

**Comment by laf** [comment]
Anyway, you guys decide what you want to do and just update this pr or close it and the other one can be looked at.

**Comment by barryodonovan** [comment]
@laf Sorry, not meaning to be snarky. I'm really impressed with the progress on librenms since I first looked at it last year. You all do really good work bringing it together. All I was saying is that of you don't accept a merge, it doesn't happen.

I know the OS concept is legacy, I just never understood it as the module definer (I guess it not actually even the OS that is being used to define it). Anyway, sorry, not meaning to be snarkey.

Barry

**Comment by guillemmateos** [comment]
Looks like the Travis CI is some sort of DNS issue with github or so. Not sure if i'll need to commit something to get it to update. Any idea @laf @murrant

**Comment by laf** [comment]
I've restarted travis build. @guillemmateos / @barryodonovan let us know when you've both decided the direction for this pr.

**Comment by guillemmateos** [comment]
I think we could merge this so we move forward and then @barryodonovan can apply any changes he considers convenient. I do think bringing barry's changes will mean a huge improvement to this module, so the faster we move, the better :)

**Comment by laf** [comment]
If the os name is wrong then there is no point in it being merged in at the moment.

**Comment by guillemmateos** [comment]
The name of the OS itself is correct. If we stick to the os name, ceragon calls it's os CeraOS. A good question though, is if, for example SAF or Cambium should then be called a different way. But if we really want to stick to the OS name, even if it's weird then Ceragon should be CeraOS.

**Comment by laf** [comment]
Existing ones can't really be changed, people could now be relying on those for alerting and possibly other items. We try and get it right at the start.

I'm not personally going to merge this until we have someone else confirm the OS name is correct.

**Comment by barryodonovan** [comment]
https://www.ceragon.com/products-ceragon/packet-hybrid-microwave/fibeair-ip-20-platform/ceraos

To be honest, the only effect that using CeraOS as the module name will have is to confuse people who look at librenms and think that it doesn't yet support ceragon. My devices will be monitored, my metrics will be graphed and my life will go on.

**Comment by laf** [comment]
@guillemmateos I can't see that you've signed the contributors, we need this before merging this code.

**Comment by guillemmateos** [comment]
@laf i'll take care of that today.

**Comment by guillemmateos** [comment]
@laf can i put the agreement on this same pull request or it needs to be on its own? My local git already has this changes and when i try to create a pull request it wants to pull also all this changes. If not, can i create a pull request just after this code is merged?

**Comment by laf** [comment]
This pull request is fine.

You really shouldn't develop against your master branch - we have docs on getting started with a recommended way if you want to take a look.

**Comment by guillemmateos** [comment]
@laf thanks

**Comment by guillemmateos** [comment]
@laf i included the contributor agreement, but it complains about conflicts on authors.md. I guess only you or somebody with write request can merge the changes

**Comment by laf** [comment]
No we can't merge until conflicts are fixed - Github disables the merge button.

This is because you need to rebase your master with ours. Should be many a tutorial online on how to do that.

**Comment by guillemmateos** [comment]
Sorry, making a bit of a mess here. trying to fix it :( my experience with github is lacking

**Comment by guillemmateos** [comment]
i think i probably close this request to avoid any serious mess. not sure what happened

**Comment by guillemmateos** [comment]
@laf i messed something with the rebase based on what i see in the pull request. i guess the easiest thing is to close this request and just create one with the right changes? sorry about it :(

**Comment by laf** [comment]
The issue is your master will now contain all of this so you're going to have to reset it back to default and replay your commits. It's possible to fix but I normally just bin my branch and port the commits over

**Comment by guillemmateos** [comment]
I'll create a new pull request with only the right changes. And i'll try to improve my github skills to avoid another mess!

**Pull request closed without merging**